velox/docs/develop/anti-join.rst

```diff
@@ -0,0 +1,581 @@
+
+.. code-block:: sql
+
+	SELECT * FROM t WHERE NOT EXISTS (
+		SELECT id FROM u WHERE u.id = t.id AND u.id IS NOT NULL
+	)
+
+This query returns 2 rows with IDs NULL and 1. In fact, the presence of NULLs in
+the subquery doesn’t affect the results of the NOT EXISTS clause. This is
+because u.id = t.id predicate evaluates to NULL when u.id is NULL, hence, rows
+with NULLs are excluded from the subquery. Unlike the NOT IN query, NOT EXISTS
+query is not sensitive for NULLs in the subquery.
+
+
+.. code-block:: sql
+
+	> WITH
+        ->     t as (SELECT * FROM unnest(array[null, 1, 2], array[0, 1, 2]) as _t(id, value)),
+        ->     u as (SELECT * FROM unnest(array[2, 3], array[1, 2]) as _t(id, value))
+        -> SELECT * from t WHERE NOT EXISTS (SELECT * FROM u WHERE u.id = t.id);
+
+	  id  | value
+	------+-------
+	    1 |     1
+	 NULL |     0
+	(2 rows)
+
+NOT EXISTS with empty subquery
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Now, consider a NOT EXISTS query with a subquery that returns empty results.
+
+.. code-block:: sql
+
+	SELECT * FROM t WHERE NOT EXISTS (
+		SELECT id FROM u WHERE u.id = t.id AND u.id < 0
+	)
+
+This query returns all rows from t because subquery always returns an empty
+result set. When subquery is empty, the results of NOT IN and NOT EXISTS
+queries are the same.
+
+.. code-block:: sql
+
+	> WITH
+        ->     t as (SELECT * FROM unnest(array[null, 1, 2], array[0, 1, 2]) as _t(id, value)),
+        ->     u as (SELECT * FROM unnest(array[], array[]) as _t(id, value))
+        -> SELECT * from t WHERE NOT EXISTS (SELECT * FROM u WHERE u.id = t.id);
+
+	  id  | value
+	------+-------
+	    2 |     2
+	    1 |     1
+	 NULL |     0
+	(3 rows)
+
+Implementation
+--------------
+
+NOT IN and NOT EXISTS queries can be implemented efficiently using anti joins.
+NOT IN queries are implemented using NULL AWARE ANTI JOIN. NOT EXISTS queries
+are implemented using regular ANTI JOIN.
+
+NULL AWARE ANTI JOIN
+~~~~~~~~~~~~~~~~~~~~
+
+NULL AWARE ANTI JOIN is used to implement NOT IN queries.
+
+.. code-block:: sql
+
+	SELECT * FROM t WHERE t.id NOT IN (SELECT id FROM u)
+
+The rows from table t are placed on the left side of the join. The rows from the
+subquery are placed on the right side of the join. The subquery rows are loaded
+into a hash table keyed on “id”. If a NULL is encountered when building the
+hash table, the join finishes early with no results. If the hash table is
+empty (i.e. subquery returns no results), the join returns all the rows from
+the left side including rows with NULL join key. If the hash table is not empty
+and has no NULLs, the rows from the left side with no NULLs in the join key are
+processed in streaming fashion. For each row, the join looks up a match in the
+hash table and returns the row only if there is no match. Rows from the left
+side with NULL in the join key are not returned.
+
+This algorithm extends trivially to multiple join keys and NOT IN queries that look like this:
```

Comment:
For multiple join keys, if one key is null, then the join key is considered as null, right?

Reply:
That's right.